### ISSUE TITLE
Allow resolve.extensions to contain [name] substitution

### ISSUE THREAD

**Issue by saiichihashimoto** [webpack/webpack#2801]
**I'm submitting a feature request**

**Webpack version:**
1.13.1

**Please tell us about your environment:**
OSX 10.11.5

**Current behavior:**
`resolve.extensions` uses the strings verbatim when resolving modules.

**Expected/desired behavior:**
`resolve.extensions` substitutes `[name]` for the entry chunk name when resolving modules.

**What is the motivation / use case for changing the behavior?**
Certain modules and components need to be used across many entry points with the only difference being that they require a different module (eg. `mymodule.entryone.js` & `mymodule.entrytwo.js`). This could include different entry points being mostly identical but wanting two different set of styles (eg. `somestyle.entryone.css` & `somestyle.entrytwo.css`)

**Browser:** all

**Language:** all

**Comment by sokra** [comment]
Try to use multiple compilations for this. 2 configs instead of 2 entry points.

**Comment by saiichihashimoto** [comment]
@sokra, that doesn't allow for using `webpack.optimize.CommonsChunkPlugin`, for example. This will end up with multiple configs with the only difference being the entry point and `resolve.extensions`. This discounts any value that the multiple entry points has. Is this a difficult change?

**Comment by sokra** [comment]
Problematic example:
A module `m` that depends on `x`.
There is `x.a.js` and `x.b.js`.
There are two entry points `a` and `b`, both depend on `m`.

When generating `m` we cannot decide between `x.a.js` / `x.b.js`. `m` need to be generated differently for `a` and `b`. This breaks the semantics of modules. Keep in mind it's also possible to run multiple entry points on a page, which would also break.

``` text
a --+        +--> x.a
    |        |
    |        ?
    |        |
    +--> m --+
    |        |
    |        ?
    |        |
b --+        +--> x.b
```

All this complex stuff where this problem is just solvable in user level. Use DI in you app. i. e. the easiest way is this:

``` js
// a.js
import { setX } from "m";
import x from "x.a";

setX(x);
bootstrap();
```

**Comment by saiichihashimoto** [comment]
This gets into a tangled mess in the use case I have. Maybe there's a correct way to approach this I'm not seeing:

- Browser extensions have different entry points:
  - content scripts (what gets injected into a specific page)
    - There could be many, many of these. Currently, I only have one.
  - background scripts (which persistently runs behind all of your sessions)
  - options script (for the options page).
  - Also, there's some other files (manifest.json, for example) that are specific to the extension as a whole.
- There will be modules that get used in the different entry points which will often be the same module, but sometimes needs a specific version of the module.
  - For example, the content script might use a module that "retrieves" a resource, but really just calls the background script and the background script actually retrieves the resource. 
- It will work in chrome, firefox, safari, etc.
- Just like the different contexts referenced in 1, there may be browser specific versions of modules. For example, an api might be different in chrome than in Firefox. Chromium and WebExtension apis mitigate most of this, but there are differences that need to be addressed.

It's possible for each module to have a set of if statements to determine which module it actually requires. However, this is specifically the case that `resolve.extensions` exists for. In this case, we would want this:

```js
// a.js

var module = require('./b');
```

With a file structure of:

```
a.js
b.js
b.background.js
b.chrome.js
b.background.firefox.js
```

The number of permutations of configurations I'd need here is absurd. With three entry points + moving the manifest.json and then (for example) three browsers, I would have to have 4 * 3 = 12 separate configs. I was going to use environment variables and use the same config for the different browser, but now I'm ALSO going to have to do the same for the entry points. Running webpack separately for each browser seems reasonable; running webpack for each combination of browser and entry point is excessive. I can't have the entry points be entry points as-is.

**TLDR** I need specific versions of some modules depending on entry point & browser. My only solutions are a tangled mess of if statements at each require or a tangled mess of an insanity of webpack configs (or one webpack config that uses env vars and is run 12 times).

How is webpack prepared for this case?
